javascript/app-discover/src/components/lists/EdgeList.utils.ts

```diff
@@ -0,0 +1,145 @@
+/*!
+ * Copyright (c) Microsoft. All rights reserved.
+ * Licensed under the MIT license. See LICENSE file in the project.
+ */
+
+import type { CausalDiscoveryConstraints } from '../../domain/CausalDiscovery/CausalDiscoveryConstraints.js'
+import type { Relationship } from '../../domain/Relationship.js'
+import {
+	hasInvertedSourceAndTarget,
+	hasSameReason,
+	hasSameSourceAndTarget,
+	invertRelationship,
+	ManualRelationshipReason,
+} from '../../domain/Relationship.js'
+
+export function removeEdge(
+	constraints: CausalDiscoveryConstraints,
+	onUpdateConstraints: (newConstraints: CausalDiscoveryConstraints) => void,
+	relationship: Relationship,
+) {
+	const newConstraints = {
+		...constraints,
+		manualRelationships: [
+			...constraints.manualRelationships.filter(
+				r =>
+					!hasSameSourceAndTarget(r, relationship) ||
+					(hasSameReason(ManualRelationshipReason.Flipped, relationship) &&
+						!hasInvertedSourceAndTarget(r, relationship)),
+			),
+			{
+				...relationship,
+				reason: ManualRelationshipReason.Removed,
+			},
+		],
+	}
+	onUpdateConstraints(newConstraints)
+}
+export function pinEdge(
+	constraints: CausalDiscoveryConstraints,
+	onUpdateConstraints: (newConstraints: CausalDiscoveryConstraints) => void,
+	relationship: Relationship,
+) {
+	const isPinned = constraints.manualRelationships.find(
+		x => x.key === relationship.key,
+	)
+	const filtered = constraints.manualRelationships.filter(
+		x => x.key !== relationship.key,
+	)
+	const newConstraints = {
+		...constraints,
+		manualRelationships: isPinned
+			? filtered
+			: [
+					...filtered,
+					{
+						...relationship,
+						reason: ManualRelationshipReason.Pinned,
+					},
+			  ],
+	}
+	onUpdateConstraints(newConstraints)
+}
+
+export function flipEdge(
+	constraints: CausalDiscoveryConstraints,
+	onUpdateConstraints: (newConstraints: CausalDiscoveryConstraints) => void,
+	closeDialogConfirm: () => void,
+	relationship?: Relationship,
+) {
+	if (!relationship) return
+	const columns = [
+		relationship.source.columnName,
+		relationship.target.columnName,
+	]
+	const reverse = invertRelationship(relationship)
+	const existingRelationships = [...constraints.manualRelationships]
+	const isFlipped = constraints.manualRelationships.find(
+		r =>
+			hasSameSourceAndTarget(r, relationship) &&
+			hasSameReason(ManualRelationshipReason.Flipped, r),
+	)
+	const filtered = existingRelationships.filter(
+		r =>
+			!(
+				hasSameSourceAndTarget(r, relationship) &&
+				hasSameReason(ManualRelationshipReason.Flipped, r)
+			),
+	)
+	const newConstraints = {
+		causes: constraints.causes.filter(r => !columns.includes(r.columnName)),
+		effects: constraints.effects.filter(r => !columns.includes(r.columnName)),
+		manualRelationships: isFlipped
```

Comment:
Same comment about logic-flipping as above - this may just be a variable name issue